File: 프로그래머스/lv3/지준혁/인사고과.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

// 1. 이중 반복문 풀이는 시간초과 (n = 100'000)
// 2. 총점으로 정렬, (순서 기억))
//    총점이 높거나 같은데 인센티브를 받을 수 없는 경우는 없다.
//    {4, 0} {5, 1}, {5, 2}, {5, 3}, {3, 4}
//    -> 정렬 {5, 1}, {5, 2}, {5, 3}, {4, 0}, {3, 4}
//    이 경우 순위는 알 수 있는데, 완호가 인센티브 제외 여부를 알 수는 없다. 어떻게 알 수 있을까?
//    가장 총점이 높은 사람만 비교해도 되지 않을까? 안됨
//    1등: {1, 7} => 8, 2등: {3, 3} => 6, 완호 {2, 0}
//    완호보다 순위가 높은 사람들 모두랑 비교해도 되지 않을까? -> 가능해보임
//	  -> 완호가 공등 1등인 경우 처리를 못함.

vector<pair<int, int>> sco;

int solution(vector<vector<int>> scores) {
    
    for (int i = 0; i < scores.size(); ++i) {
        int sum = 0;
        for (auto e : scores[i]) {
            sum += e;
        }
        sco.push_back({sum, i});
    }
    
  	sort(sco.begin(), sco.end(), [](auto& a, auto& b) {
        return a > b;
    });    
    
    int rank = 2;
    int wanho_rank = 1;
    for (int i = 0; i < sco.size(); ++i) {
        cout << sco[i].first << ' ' << sco[i].second << '\n';
        if (sco[0].second == 0) break;
        if (i == 0) continue;
        if (sco[i].first < sco[i - 1].first) // 이전사람이 점수가 더 높다면 등수 올림
            ++rank;
        if (sco[i].second == 0) {
            wanho_rank = rank;
        }
    }
    
    bool is_not_incentive = false;
    for (int i = 0; i < rank; ++i) {
        int index = sco[i].second;
        if (scores[0][0] < scores[index][0] && scores[0][1] < scores[index][1]) {
            is_not_incentive = true;
        }
    }
    if (is_not_incentive) {
        return -1;
    }
    return rank;
}

File: 프로그래머스/lv3/최원준/인사고과.py
#

'''
1. 아이디어 :
    내림차순, 오름차순 정렬한다.
2. 시간복잡도 :
    O(nlogn)
3. 자료구조 :
    배열
'''

def solution(scores):
    target = scores[0]
    target_sum = sum(target)
    scores.sort(key=lambda x: (-x[0], x[1])) #정렬을 통해, 항상 뒤에 있는
    # 	[[3, 2], [3, 2], [2, 1], [2, 2], [1, 4]]

    cmax = 0
    ans = 1
    for score in scores:
        if target[0] < score[0] and target[1] < score[1]:
            return -1
        if cmax <= score[1]: #cmax보다 작으면
            if target_sum < score[0] + score[1]:
                ans += 1
            cmax = score[1]
            #항상 1번째 값은 최대값으로 한다.
            #뒤에 오는 값들은 0번의 값으로 정렬되어있으므로,
            #cmax보다 크거나 같다면, 지워지지 않는 거고,
            #지워지지 않으면 target_sum과 비교해서 랭크를 하나씩 밀어낸다.
    return ans
